### PR TITLE
revise cachedataset runtime_cache modes

### LINKED ISSUE #5613: CacheDataset shared memory,  during torchrun multi-gpu training
CacheDataset with shared memory crashes during torchrun distributed training

steps to reproduce, run `torchrun --nproc_per_node=2 main.py`

```
import torch, os
import torch.distributed as dist
import torch.multiprocessing as mp
from monai.data import CacheDataset


rank = int(os.getenv("LOCAL_RANK"))
dist.init_process_group(backend="nccl", init_method="env://") 
torch.cuda.set_device(rank)

dataset = CacheDataset(data=[1,2,3], runtime_cache=True)
```

this errors doesn't accure if spawning multi-gpu processes manually (instead of torchrun). 
this issue is  because "broadcast_object_list" checks for authkey in each process, and  torchrun has these keys different.

### PR DESCRIPTION
**Pull request by wyli**
Fixes #5613

### Description

- `runtime_cache=False`: the default v1.0.1 behaviour
- `runtime_cache=True` or `"thread"`: single process, for caching cuda tensors
- `runtime_cache="process"`: single process workflow + multiprocess dataloader
- `runtime_cache=` user-provided object, could be used to pass a container shared among processes

I feel in this way the user can determine what to use instead of guessing and providing an automated solution... let me know what you think @myron @Nic-Ma, I'm fine if this is eventually merged or not merged

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/data/dataset.py [hunk 1]
```diff
@@ -777,18 +776,21 @@ def __init__(
...
-                `monai.data.DataLoader`, and can't work with `monai.data.DistributedSampler`
-                as GPU Tensor usually can't be shared in the multiprocessing context.
-                (try ``cache_dataset.disable_share_memory_cache()`` in case of GPU caching issues.)
+            runtime_cache: mode of cache at the runtime. Default to `False` to prepare
+                the cache content for the entire ``data`` during initialization, this potentially largely increase the
+                time required between the constructor called and first mini-batch generated.
+                Three options are provided to compute the cache on the fly after the dataset initialization:
 
+                1. ``"thread"`` or ``True``: use a regular ``list`` to store the cache items.
+                2. ``"process"``: use a ListProxy to store the cache items, it can be shared among processes.
```
Comment (Nic-Ma) [inline monai/data/dataset.py:785]: I would suggest to use "multithread" and "multiprocessing" words.

Thanks.

File: monai/data/dataset.py [hunk 2]
```diff
@@ -777,18 +776,21 @@ def __init__(
...
+                3. A list-like object: a users-provided container to be used to store the cache items.
+
+                For `thread-based` cache (typically for caching cuda tensors), option 1 is recommended.
+                For single process workflow with multiprocess data loading, option 2 is recommended.
```
Comment (Nic-Ma) [inline monai/data/dataset.py:789]: multiprocess -> multiprocessing

File: monai/data/dataset.py [hunk 3]
```diff
@@ -777,18 +776,21 @@ def __init__(
...
+                For multiprocess workflow (typically for distributed training),
```
Comment (Nic-Ma) [inline monai/data/dataset.py:790]: multiprocess -> multiprocessing

**Review by Nic-Ma — APPROVED** [review]

**Comment by Nic-Ma** [comment]
The overall solution looks good to me and it matches the option 2 from @myron in https://github.com/Project-MONAI/MONAI/pull/5615#discussion_r1037904822

Thanks.

**Comment by myron** [comment]
This PR was created and merged without a discussion. Because it was done overnight, It started at 3am PST and merged at 10am PST, it did not give an opportunity for a review or a discussion.
We spent  weeks  working on an easy shared cache support, https://github.com/Project-MONAI/MONAI/pull/5365, https://github.com/Project-MONAI/MONAI/pull/5308.  Most recently we spend over 10 days discussion  this  small PR https://github.com/Project-MONAI/MONAI/pull/5565 which was heavily scrutinized, and as soon as it was finally merged, the current PR (https://github.com/Project-MONAI/MONAI/pull/5630) removed all of that effort.  

@wyli  I don't believe it is a good project MONAI practice, you should have a meaningful discussion about it first.

**Pull request merged**